File: trainer-flutter/lib/models/training.dart
class Training {
  final int id;
  final String? title;
  final DateTime? startTime;
  final DateTime? endTime;
  final String? date;
  final String? timeFrom;
  final String? timeTo;
  final int? clientId;
  final String? clientName;
  final int? trainerId;
  final String? trainerName;
  final int? locationId;
  final String? locationName;
  final String? trainingType;
  final String? status;
  final String? notes;
  final bool isCancelled;

  const Training({
    required this.id,
    this.title,
    this.startTime,
    this.endTime,
    this.date,
    this.timeFrom,
    this.timeTo,
    this.clientId,
    this.clientName,
    this.trainerId,
    this.trainerName,
    this.locationId,
    this.locationName,
    this.trainingType,
    this.status,
    this.notes,
    this.isCancelled = false,
  });

  factory Training.fromJson(Map<String, dynamic> json) {
    DateTime? startTime;
    DateTime? endTime;

    // Try combined datetime fields first
    final startStr = json['start']?.toString() ??
        json['start_time']?.toString() ??
        json['datetime_from']?.toString();
    final endStr = json['end']?.toString() ??
        json['end_time']?.toString() ??
        json['datetime_to']?.toString();

    if (startStr != null) startTime = DateTime.tryParse(startStr);
    if (endStr != null)   endTime   = DateTime.tryParse(endStr);

    // Fallback: date + separate time field
    // NestJS uses "starttime" (no underscore); PHP uses "time_from" / "from"
    if (startTime == null) {
      final date = json['date']?.toString() ?? json['training_date']?.toString();
      final timeFrom = json['time_from']?.toString() ??
          json['starttime']?.toString() ??
          json['from']?.toString() ??
          json['start_time']?.toString();
      if (date != null && timeFrom != null) {
        startTime = DateTime.tryParse('$date $timeFrom');
      }
    }

    // NestJS embeds full client/trainer objects; PHP uses flat fields
    final clientObj   = json['client']   as Map<String, dynamic>?;
    final trainerObj  = json['trainer']  as Map<String, dynamic>?;

    String? clientName = json['client_name']?.toString() ??
        _buildName(json, 'client_first_name', 'client_last_name');
    if (clientName == null && clientObj != null) {
      clientName = _buildName(clientObj, 'firstname', 'lastname');
    }

    String? trainerName = json['trainer_name']?.toString();
    if (trainerName == null && trainerObj != null) {
      trainerName = _buildName(trainerObj, 'firstname', 'lastname');
    }

    final statusStr =
        json['status']?.toString() ?? json['training_status']?.toString();

    return Training(
      id: _parseInt(json['id'] ?? json['training_id'] ?? 0),
      title: json['title']?.toString() ?? json['name']?.toString(),
      startTime: startTime,
      endTime: endTime,
      date: json['date']?.toString() ?? json['training_date']?.toString(),
      // NestJS "starttime"; PHP "time_from"
      timeFrom: json['time_from']?.toString() ??
          json['starttime']?.toString() ??
          json['from']?.toString(),
      timeTo: json['time_to']?.toString() ?? json['to']?.toString(),
      // NestJS camelCase or PHP snake_case
      clientId:  _parseInt(json['client_id']  ?? json['clientId']),
      clientName: clientName,
      trainerId: _parseInt(json['trainer_id'] ?? json['trainerId']),
      trainerName: trainerName,
      locationId: _parseInt(json['location_id'] ?? json['locationId']),
      locationName: json['location_name']?.toString() ??
          json['location']?.toString(),
      trainingType: json['training_type']?.toString() ??
          json['type']?.toString() ??
          json['type_name']?.toString(),
      status: statusStr,
      notes: json['notes']?.toString() ?? json['comment']?.toString(),
      isCancelled: statusStr?.toLowerCase() == 'cancelled' ||
          statusStr?.toLowerCase() == 'canceled' ||
          (json['is_cancelled'] == true) ||
          (json['cancelled'] == 1) ||
          (json['cancelled'] == true),
    );
  }

  static String? _buildName(Map<String, dynamic> json, String firstKey, String lastKey) {
    final first = json[firstKey]?.toString() ?? '';
    final last = json[lastKey]?.toString() ?? '';
    final name = '$first $last'.trim();
    return name.isNotEmpty ? name : null;
  }

  static int _parseInt(dynamic value) {
    if (value == null) return 0;
    if (value is int) return value;
    if (value is String) return int.tryParse(value) ?? 0;
    return 0;
  }

  String get displayDate {
    if (startTime != null) {
      return '${startTime!.day.toString().padLeft(2, '0')}.${startTime!.month.toString().padLeft(2, '0')}.${startTime!.year}';
    }
    return date ?? 'Unknown date';
  }

  String get displayTime {
    if (startTime != null) {
      final start =
          '${startTime!.hour.toString().padLeft(2, '0')}:${startTime!.minute.toString().padLeft(2, '0')}';
      if (endTime != null) {
        final end =
            '${endTime!.hour.toString().padLeft(2, '0')}:${endTime!.minute.toString().padLeft(2, '0')}';
        return '$start - $end';
      }
      return start;
    }
    if (timeFrom != null) {
      if (timeTo != null) return '$timeFrom - $timeTo';
      return timeFrom!;
    }
    return 'Unknown time';
  }
}

File: trainer-ios/SihlTrainer/Models/Training.swift
import Foundation

/// Pendant zu `models/training.dart` — ein Termin/Training im Kalender.
struct Training: Identifiable, Equatable {
    let id: Int
    let title: String?
    let startTime: Date?
    let endTime: Date?
    let day: String?
    let timeFrom: String?
    let timeTo: String?
    let clientId: Int?
    let clientName: String?
    let trainerId: Int?
    let trainerName: String?
    let locationId: Int?
    let locationName: String?
    let trainingType: String?
    let status: String?
    let notes: String?
    let isCancelled: Bool
    let cancelledAt: Date?
    let trainingPlanId: Int?
    let trainingPlanName: String?

    init(json: [String: Any]) {
        id = JSON.int(json, "id", "training_id") ?? 0
        title = JSON.string(json, "title", "name")

        // Erst kombinierte Datetime-Felder, dann date + getrennte Zeit
        // (NestJS: "starttime" ohne Unterstrich, PHP: "time_from"/"from").
        let start = JSON.date(JSON.string(json, "start", "start_time", "datetime_from"))
        let dayValue = JSON.string(json, "date", "training_date")
        let from = JSON.string(json, "time_from", "starttime", "from", "start_time")
        startTime = start ?? JSON.date(day: dayValue, time: from)
        endTime = JSON.date(JSON.string(json, "end", "end_time", "datetime_to"))

        day = dayValue
        timeFrom = JSON.string(json, "time_from", "starttime", "from")
        timeTo = JSON.string(json, "time_to", "to")

        // NestJS bettet client/trainer als Objekt ein, PHP liefert flache Felder.
        let clientObject = json["client"] as? [String: Any]
        let trainerObject = json["trainer"] as? [String: Any]

        clientId = JSON.intNonZero(json, "client_id", "clientId")
        if let name = JSON.string(json, "client_name") {
            clientName = name
        } else if let clientObject {
            clientName = Training.name(from: clientObject)
        } else {
            clientName = Training.name(from: json, first: "client_first_name", last: "client_last_name")
        }

        trainerId = JSON.intNonZero(json, "trainer_id", "trainerId")
        if let name = JSON.string(json, "trainer_name") {
            trainerName = name
        } else if let trainerObject {
            trainerName = Training.name(from: trainerObject)
        } else {
            trainerName = nil
        }

        locationId = JSON.intNonZero(json, "location_id", "locationId")
        locationName = JSON.string(json, "location_name", "location")
        trainingType = JSON.string(json, "training_type", "trainingType", "type")
        status = JSON.string(json, "status", "training_status")
        notes = JSON.string(json, "notes", "comment")
        isCancelled = JSON.bool(json, "cancelled", "is_cancelled", "isCancelled")
            || (JSON.string(json, "status", "training_status")?.lowercased() == "cancelled")
        cancelledAt = JSON.date(JSON.string(json, "cancelled_at", "cancelledAt"))
        trainingPlanId = JSON.intNonZero(json, "training_plan_id", "trainingplan_id", "trainingPlanId")
        trainingPlanName = JSON.string(json, "training_plan_name", "trainingplan_name")
    }

    private static func name(from json: [String: Any],
                             first: String = "firstname",
                             last: String = "lastname") -> String? {
        let firstName = JSON.string(json, first, "first_name", "name") ?? ""
        let lastName = JSON.string(json, last, "last_name", "surname") ?? ""
        let full = "\(firstName) \(lastName)".trimmingCharacters(in: .whitespaces)
        return full.isEmpty ? nil : full
    }

    /// Kurzfristige Absage: weniger als 12 Stunden vor Beginn abgesagt. Diese
    /// Termine werden weiterhin verrechnet und bleiben im Kalender sichtbar.
    var isLateCancellation: Bool {
        guard isCancelled, let startTime, let cancelledAt else { return false }
        return startTime.timeIntervalSince(cancelledAt) < 12 * 3600
    }
}
